{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Metage2Metabo\n",
    "\n",
    "Metage2Metabo (m2m) is a python package to perform graph-based metabolic analysis starting from annotated genomes or metabolic networks.\n",
    "\n",
    "This jupyter notebook uses the following versions of these dependencies:\n",
    "\n",
    "- Metage2Metabo==1.6.0\n",
    "- mpwt==0.6.0\n",
    "- padmet==5.0.1\n",
    "- MeneTools==3.4.0\n",
    "- Miscoto==3.2.0\n",
    "- bubbletools==0.6.11\n",
    "- networkx==2.5\n",
    "- pandas\n",
    "- tabulate==0.9.0"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Metage2Metabo: 1.5.4\n",
      "padmet: 5.0.1\n",
      "MeneTools: 3.3.0\n",
      "Miscoto: 3.1.2\n",
      "networkx: 0.6.11\n",
      "networkx: 3.2.1\n",
      "pandas: 2.1.4\n",
      "tabulate: 0.9.0\n"
     ]
    }
   ],
   "source": [
    "from metage2metabo import __version__ as m2m_version\n",
    "from menetools import __version__ as menetools_version\n",
    "from miscoto import __version__ as miscoto_version\n",
    "from networkx import __version__ as networkx_version\n",
    "from pandas import __version__ as pandas_version\n",
    "from tabulate import __version__ as tabulate_version\n",
    "\n",
    "from importlib.metadata import version\n",
    "padmet_version = version('padmet')\n",
    "bubbletools_version = version('bubbletools')\n",
    "\n",
    "print('Metage2Metabo: ' + m2m_version)\n",
    "print('padmet: ' + padmet_version)\n",
    "print('MeneTools: ' + menetools_version)\n",
    "print('Miscoto: ' + miscoto_version)\n",
    "print('networkx: ' + bubbletools_version)\n",
    "print('networkx: ' + networkx_version)\n",
    "print('pandas: ' + pandas_version)\n",
    "print('tabulate: ' + tabulate_version)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## `m2m iscope`: Individual scope\n",
    "\n",
    "This part of the workflow encompasses the computation of the individual scope.\n",
    "\n",
    "### Scope formalisation\n",
    "\n",
    "We define a metabolic network as a bipartite graph $G = (R \\cup M, E)$, where $R$ and $M$ stand for reaction and metabolite nodes.  When $(m,r) \\in E$ (respectively $(r,m) \\in E$), with $m \\in M$  and $r \\in R$, the metabolite is called a _reactant_ (respectively _product_) of the reaction $r$. The scope of a set of seed compounds $S$ according to a metabolic network $G$, denoted by $Scope(G, S)$, is iteratively computed until it reaches a fixed point (Handorf and Ebenhöh 2005). It is formally defined by:\n",
    "\n",
    "$$\\Scope(G, S) = \\bigcup_iM_i, \\mbox{ where } M_0 = S \\mbox{ and }\n",
    "M_{i+1} = M_i \\cup \\Products{\\{ r \\in R \\mid \\Reactants\\{r\\} \\subseteq M_i\\}}.$$\n",
    "\n",
    "### An example\n",
    "\n",
    "In the following gif, we look at a metabolic network with 3 reactions (R1, R2, R3) and 6 metabolites (A, B, C, D, E, F).\n",
    "\n",
    "The seeds are composed of metabolites A and B.\n",
    "\n",
    "With these inputs, the scope will be computed. As A and B are available as seed, they can be used as reactant to activate the reaction R1, which will produce metabolite D. But the reaction R2 will not be activated as the metabolite C is not availabe in the seeds.\n",
    "\n",
    "The metabolite D, produced by the reaction R1, can activate the reaction R3 to produce the metabolite F.\n",
    "\n",
    "<p>\n",
    "    <img src='images/iscope.gif' width='30%'>\n",
    "    <center><em>Figure 1: Individual scope</em></center>\n",
    "</p>\n",
    "\n",
    "### Tutorial\n",
    "\n",
    "For this tutorial, we will use an example of a community consisting of 6 bacteria, i.e. 6 metabolic networks. The medium used for the simulation consists of 2 metabolites, $S_1$ and $S_2$. Additionally, we are interested in the production of 4 metabolites, $H$, $C$, $E$ and $foo$. In the SBML files of the dataset, each metabolite has the prefix $M\\_$, and a suffix corresponding to the compartment it belongs to, e.g. \"M_S1_c\" is the metabolite $S_1$ in the cytosol. \n",
    "\n",
    "The metabolic networks of `bact1`, `bact2`, `bact3`, `bact4`, `bact5` and `bact6` are depicted below:\n",
    "\n",
    "<p>\n",
    "    <img src='images/tiny_toy_networks.png' width='70%'>\n",
    "    <center><em>Figure 2: Metabolic networks of the toy community</em></center>\n",
    "</p>\n",
    "\n",
    "In each of the networks, the compounds of the nutritional environmenent (i.e. the seeds) are depicted in yellow, and the metabolites we are interested in (i.e. the targets) are depicted in green.\n",
    "\n",
    "Let's look at the individual scope of each metabolic network in the medium consisting of the two seeds. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'M_N_c', 'M_D_c', 'M_H_c', 'M_A_c', 'M_R_c', 'M_S2_c', 'M_B_c'}\n"
     ]
    }
   ],
   "source": [
    "# Instanciate the logger\n",
    "import logging\n",
    "\n",
    "logging.basicConfig()\n",
    "\n",
    "# Run the individual scope with metage2metabo on the folder of SBML using the seeds files.\n",
    "# This is similar to the command:\n",
    "# m2m iscope -n data/community -s data/seeds.sbml -o output_folder\n",
    "from metage2metabo.m2m import individual_scope\n",
    "\n",
    "union_of_iscopes = individual_scope.iscope('../../test/metabolic_data/tiny_toy/networks/', '../../test/metabolic_data/tiny_toy/seeds_community.sbml', 'output_folder')\n",
    "\n",
    "## The scopes for each species will be computed in the output_folder. The union of all individual scopes will be returned as a set. \n",
    "\n",
    "print(union_of_iscopes)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The union of the individual scope encompasses 7 metabolites. More details are provided in the output files. Let's look at them."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Individual scope:\n",
      "bact1: 5 metabolites (M_A_c, M_B_c, M_D_c, M_H_c, M_R_c).\n"
     ]
    }
   ],
   "source": [
    "# Read the result\n",
    "\n",
    "import json\n",
    "\n",
    "with open('output_folder/indiv_scopes/indiv_scopes.json') as iscope_output:\n",
    "    iscope_json_data = json.load(iscope_output)\n",
    "\n",
    "print('Individual scope:')\n",
    "\n",
    "print('{0}: {1} metabolites ({2}).'.format('bact1', len(iscope_json_data['bact1']), ', '.join(sorted(iscope_json_data['bact1']))))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We can check the individual scope of each network (circled in orange below):\n",
    "\n",
    "\n",
    "<p>\n",
    "    <img src='images/tiny_toy_iscopes.png' width='70%'>\n",
    "    <center><em>Figure 3: Individual scopes</em></center>\n",
    "</p>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "bact5: 0 metabolites ().\n",
      "bact3: 2 metabolites (M_A_c, M_B_c).\n",
      "bact6: 0 metabolites ().\n",
      "bact2: 4 metabolites (M_A_c, M_B_c, M_N_c, M_S2_c).\n",
      "bact4: 1 metabolites (M_B_c).\n",
      "bact1: 5 metabolites (M_A_c, M_B_c, M_D_c, M_H_c, M_R_c).\n"
     ]
    }
   ],
   "source": [
    "for org in iscope_json_data.keys():\n",
    "    print('{0}: {1} metabolites ({2}).'.format(org, len(iscope_json_data[org]), ', '.join(sorted(iscope_json_data[org]))))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We notice that two species cannot produce anything from the seeds: `bact5` and `bact6`. \n",
    "\n",
    "Species `bact1` has the largest scope, 5 metabolites. \n",
    "\n",
    "Species `bact2` has the potential to produce one of the seeds, $S_2$. \n",
    "\n",
    "We notice that contrary to the initial definition of the scope above, only the seeds that can really be produced, i.e are the product of a reaction that is predicted to be activated, are included in the scopes. \n",
    "\n",
    "More generally, one might be interested in checking the status of the seeds, and take a look at the dedicated output:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Seeds that cannot be individually produced:\n",
      "{'bact1': ['M_S2_c', 'M_S1_c'], 'bact2': ['M_S1_c'], 'bact3': ['M_S2_c', 'M_S1_c'], 'bact4': ['M_S2_c'], 'bact5': ['M_S1_c'], 'bact6': ['M_S1_c']}\n",
      "\n",
      "Seeds that can be individually produced:\n",
      "{'bact1': [], 'bact2': ['M_S2_c'], 'bact3': [], 'bact4': [], 'bact5': [], 'bact6': []}\n",
      "\n",
      "Seeds that do not appear in the metabolic networks:\n",
      "{'bact1': [], 'bact2': [], 'bact3': [], 'bact4': ['M_S1_c'], 'bact5': ['M_S2_c'], 'bact6': ['M_S2_c']}\n"
     ]
    }
   ],
   "source": [
    "with open('output_folder/indiv_scopes/seeds_in_indiv_scopes.json') as f:\n",
    "    iscope_seeds_data = json.load(f)\n",
    "\n",
    "print('Seeds that cannot be individually produced:')\n",
    "print(iscope_seeds_data['individually_non_producible_seeds'])\n",
    "\n",
    "print('\\nSeeds that can be individually produced:')\n",
    "print(iscope_seeds_data['individually_producible_seeds'])\n",
    "\n",
    "print('\\nSeeds that do not appear in the metabolic networks:')\n",
    "print(iscope_seeds_data['seeds_absent_in_metabolic_network'])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Notable information here is that $S_1$ does not appear in `bact2` and $S_2$ does not appear in `bact5` nor `bact6`. A context in which a seed never occurs in any of the networks might indicate an error in the seed definition, a typo for instance. \n",
    "\n",
    "The fact that a species has the ability to produce a seed could be interesting to suggest a renewal of the seed, and less competition for that molecule. \n",
    "\n",
    "Other outputs related to the individual scope computation are the reversed iscopes. They provide the same information but with a focus on the molecules rather than the species: which species are predicted to produce each molecule? The information is stored as a json file and as a matrix."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{\n",
      "    \"M_A_c\": [\n",
      "        \"bact3\",\n",
      "        \"bact2\",\n",
      "        \"bact1\"\n",
      "    ],\n",
      "    \"M_B_c\": [\n",
      "        \"bact3\",\n",
      "        \"bact2\",\n",
      "        \"bact4\",\n",
      "        \"bact1\"\n",
      "    ],\n",
      "    \"M_D_c\": [\n",
      "        \"bact1\"\n",
      "    ],\n",
      "    \"M_H_c\": [\n",
      "        \"bact1\"\n",
      "    ],\n",
      "    \"M_N_c\": [\n",
      "        \"bact2\"\n",
      "    ],\n",
      "    \"M_R_c\": [\n",
      "        \"bact1\"\n",
      "    ],\n",
      "    \"M_S2_c\": [\n",
      "        \"bact2\"\n",
      "    ]\n",
      "}\n"
     ]
    }
   ],
   "source": [
    "with open('output_folder/indiv_scopes/rev_iscope.json') as f:\n",
    "    rev_iscope = json.load(f)\n",
    "\n",
    "print(json.dumps(rev_iscope, indent = 4, sort_keys=True))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>M_A_c</th>\n",
       "      <th>M_B_c</th>\n",
       "      <th>M_N_c</th>\n",
       "      <th>M_S2_c</th>\n",
       "      <th>M_D_c</th>\n",
       "      <th>M_H_c</th>\n",
       "      <th>M_R_c</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>bact5</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>bact3</th>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>bact6</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>bact2</th>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>bact4</th>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>bact1</th>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       M_A_c  M_B_c  M_N_c  M_S2_c  M_D_c  M_H_c  M_R_c\n",
       "bact5      0      0      0       0      0      0      0\n",
       "bact3      1      1      0       0      0      0      0\n",
       "bact6      0      0      0       0      0      0      0\n",
       "bact2      1      1      1       1      0      0      0\n",
       "bact4      0      1      0       0      0      0      0\n",
       "bact1      1      1      0       0      1      1      1"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "import pandas \n",
    "import csv\n",
    "from io import StringIO\n",
    "\n",
    "display(pandas.read_csv('output_folder/indiv_scopes/rev_iscope.tsv', sep='\\t', index_col='Unnamed: 0'))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Command line from the terminal\n",
    "\n",
    "Most of the time, you will launch m2m from the terminal rather than from the API. \n",
    "The command for individual scope computation is the following, it will create all output files:\n",
    "\n",
    "```sh\n",
    "m2m iscope -n ../../test/metabolic_data/tiny_toy/networks/ -s ../../test/metabolic_data/tiny_toy/seeds_community.sbml -o output_folder\n",
    "```\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## `m2m cscope`: Community scope\n",
    "\n",
    "This part encompasses the computation of the community scope, i.e. the metabolic potential of the community considering that interactions could occur between members. By interactions, we hereby mean cross feedings, or metabolic exchanges. We consider that any molecule can become a by-product or be shared through cross-feeding to other bacterial populations. We therefore assess the upper bound of the metabolic potential, considering that it is unlikely that all metabolites can be shared to all species. \n",
    "\n",
    "### Community scope formalisation\n",
    "\n",
    "The `m2m cscope` command computes the metabolic capabilities of the whole microbiota by taking into account the complementarity of metabolic pathways between metabolic networks. This step simulates the sharing of metabolic biosynthesis through a meta-organism composed of all metabolic networks, and assesses the metabolic compounds that can be reached using network expansion. This calculation is an extension of the features of [MiSCoTo](https://github.com/cfrioux/miscoto) in which the collective scope of a collection of metabolic networks $\\{G_1, \\dots G_N\\}$ is introduced. We define:\n",
    "\n",
    "$$CommunityScope(G_1.. G_N, S)=Scope \\left( \\left(\\bigcup_{i \\in {\\{1..n\\}}} R_i, \\bigcup_{i \\in {\\{1..n\\}}} M_i, \\bigcup_{i \\in {\\{1..n\\}}} E_i \\right), S \\right).$$ \n",
    "\n",
    "### Application to the example dataset. \n",
    "\n",
    "In this example, we have 6 metabolic networks (`bact1`, `bact2`, `bact3`, `bact4`, `bact5` and `bact6`). As for the Individual Scope, the seeds are composed of two metabolites (S1 and S2) and highlighted in yellow. With these inputs, the Community Scope can be computed.\n",
    "\n",
    "The formalism behind the community scope of m2m uses a Mixed Bag modelling, which considers the metabolic networks of the community as a meta-organism allowing exchanges between them without cost. For example, no reaction in the metabolic network `bact2` produces metabolite `C` as reactant but reactions of metabolic networks `bact5` and `bact6` does. With the Mixed Bag modelling, the metabolite `C` is then available to be used by `bact2` after being produced by `bact5` and `bact6`. \n",
    "\n",
    "\n",
    "The Community Scope of the 6 organisms is composed of of 15 metabolites, all circled in blue.\n",
    "\n",
    "<p>\n",
    "    <img src='images/tiny_toy_cscope.png' width='60%'>\n",
    "    <center><em>Figure 4: Community scope</em></center>\n",
    "</p>\n",
    "\n",
    "We could visualise the meta-metabolism of an organism with all the merged reactions of the 6 species. \n",
    "\n",
    "<p>\n",
    "    <img src='images/tiny_toy_cscope_metaorg.png' width='50%'>\n",
    "    <center><em>Figure 5: Corresponding meta-organism sharing all metabolic capabilities</em></center>\n",
    "</p>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The community scope contains 15 metabolites: M_A_c, M_B_c, M_C_c, M_D_c, M_E_c, M_F_c, M_G_c, M_H_c, M_K_c, M_N_c, M_R_c, M_S1_c, M_S2_c, M_V_c, M_X_c.\n"
     ]
    }
   ],
   "source": [
    "# Run the community scope with metage2metabo on the folder of SBML using the seeds files.\n",
    "# This is similar to the command:\n",
    "#m2m cscope -n data/community -s data/seeds.sbml -o output_folder\n",
    "from metage2metabo.m2m import community_scope\n",
    "\n",
    "instance_path, network_scopes = community_scope.cscope('../../test/metabolic_data/tiny_toy/networks/', '../../test/metabolic_data/tiny_toy/seeds_community.sbml', 'output_folder')\n",
    "\n",
    "print('The community scope contains {0} metabolites: {1}.'.format(len(network_scopes), ', '.join(sorted(network_scopes))))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "It is however important to know which species are responsible for the prodicibility of each molecule, i.e. has an activated reaction that produce the molecules (although the activation of this reaction might rely on interactions with other species). \n",
    "\n",
    "This information is also provided by `m2m cscope`, in a dedicated output file. \n",
    "\n",
    "Overall the community scope command creates 4 output files : \n",
    "\n",
    "- the list of the compounds in the community scope, stored in `output_folder/community_analysis/comm_scopes.json`\n",
    "- the contribution of each microbe to the community scope in `output_folder/community_analysis/contributions_of_microbes.json`\n",
    "- the reverse community scope, with a focus on the molecules rather that the species, in json format\n",
    "- the reverse community scope, with a focus on the molecules rather that the species, in tabulated format\n",
    "\n",
    "Let's look at the outputs:\n",
    "\n",
    "First the community scope itself:\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{\n",
      "    \"com_scope\": [\n",
      "        \"M_N_c\",\n",
      "        \"M_S1_c\",\n",
      "        \"M_D_c\",\n",
      "        \"M_V_c\",\n",
      "        \"M_C_c\",\n",
      "        \"M_H_c\",\n",
      "        \"M_A_c\",\n",
      "        \"M_K_c\",\n",
      "        \"M_G_c\",\n",
      "        \"M_E_c\",\n",
      "        \"M_X_c\",\n",
      "        \"M_F_c\",\n",
      "        \"M_R_c\",\n",
      "        \"M_S2_c\",\n",
      "        \"M_B_c\"\n",
      "    ]\n",
      "}\n"
     ]
    }
   ],
   "source": [
    "with open('output_folder/community_analysis/comm_scopes.json') as f:\n",
    "    cscope = json.load(f)\n",
    "\n",
    "print(json.dumps(cscope, indent = 4, sort_keys=True))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The contribution of microbes file describes what each community member produces individually (iscope), and in community, and highlights the added-value of interactions, i.e. what can be produced in community that cannot be produced individually."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{\n",
      "    \"bact1\": {\n",
      "        \"community_metabolic_gain\": [\n",
      "            \"M_F_c\"\n",
      "        ],\n",
      "        \"produced_alone\": [\n",
      "            \"M_R_c\",\n",
      "            \"M_H_c\",\n",
      "            \"M_A_c\",\n",
      "            \"M_D_c\",\n",
      "            \"M_B_c\"\n",
      "        ],\n",
      "        \"produced_in_community\": [\n",
      "            \"M_H_c\",\n",
      "            \"M_A_c\",\n",
      "            \"M_B_c\",\n",
      "            \"M_R_c\",\n",
      "            \"M_D_c\",\n",
      "            \"M_F_c\"\n",
      "        ]\n",
      "    },\n",
      "    \"bact2\": {\n",
      "        \"community_metabolic_gain\": [\n",
      "            \"M_H_c\",\n",
      "            \"M_G_c\",\n",
      "            \"M_E_c\"\n",
      "        ],\n",
      "        \"produced_alone\": [\n",
      "            \"M_B_c\",\n",
      "            \"M_S2_c\",\n",
      "            \"M_N_c\",\n",
      "            \"M_A_c\"\n",
      "        ],\n",
      "        \"produced_in_community\": [\n",
      "            \"M_G_c\",\n",
      "            \"M_H_c\",\n",
      "            \"M_A_c\",\n",
      "            \"M_E_c\",\n",
      "            \"M_B_c\",\n",
      "            \"M_S2_c\",\n",
      "            \"M_N_c\"\n",
      "        ]\n",
      "    },\n",
      "    \"bact3\": {\n",
      "        \"community_metabolic_gain\": [\n",
      "            \"M_N_c\",\n",
      "            \"M_S1_c\",\n",
      "            \"M_H_c\",\n",
      "            \"M_E_c\",\n",
      "            \"M_G_c\"\n",
      "        ],\n",
      "        \"produced_alone\": [\n",
      "            \"M_B_c\",\n",
      "            \"M_A_c\"\n",
      "        ],\n",
      "        \"produced_in_community\": [\n",
      "            \"M_S1_c\",\n",
      "            \"M_B_c\",\n",
      "            \"M_N_c\",\n",
      "            \"M_G_c\",\n",
      "            \"M_H_c\",\n",
      "            \"M_A_c\",\n",
      "            \"M_E_c\"\n",
      "        ]\n",
      "    },\n",
      "    \"bact4\": {\n",
      "        \"community_metabolic_gain\": [\n",
      "            \"M_X_c\"\n",
      "        ],\n",
      "        \"produced_alone\": [\n",
      "            \"M_B_c\"\n",
      "        ],\n",
      "        \"produced_in_community\": [\n",
      "            \"M_B_c\",\n",
      "            \"M_X_c\"\n",
      "        ]\n",
      "    },\n",
      "    \"bact5\": {\n",
      "        \"community_metabolic_gain\": [\n",
      "            \"M_V_c\",\n",
      "            \"M_C_c\",\n",
      "            \"M_K_c\"\n",
      "        ],\n",
      "        \"produced_alone\": [],\n",
      "        \"produced_in_community\": [\n",
      "            \"M_K_c\",\n",
      "            \"M_V_c\",\n",
      "            \"M_C_c\"\n",
      "        ]\n",
      "    },\n",
      "    \"bact6\": {\n",
      "        \"community_metabolic_gain\": [\n",
      "            \"M_X_c\",\n",
      "            \"M_C_c\",\n",
      "            \"M_K_c\"\n",
      "        ],\n",
      "        \"produced_alone\": [],\n",
      "        \"produced_in_community\": [\n",
      "            \"M_X_c\",\n",
      "            \"M_C_c\",\n",
      "            \"M_K_c\"\n",
      "        ]\n",
      "    }\n",
      "}\n"
     ]
    }
   ],
   "source": [
    "with open('output_folder/community_analysis/contributions_of_microbes.json') as f:\n",
    "    contribs = json.load(f)\n",
    "\n",
    "print(json.dumps(contribs, indent = 4, sort_keys=True))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The reverse community scopes is a dictionnary with compounds as keys and species that produce them in community as values."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{\n",
      "    \"M_A_c\": [\n",
      "        \"bact3\",\n",
      "        \"bact2\",\n",
      "        \"bact1\"\n",
      "    ],\n",
      "    \"M_B_c\": [\n",
      "        \"bact3\",\n",
      "        \"bact2\",\n",
      "        \"bact4\",\n",
      "        \"bact1\"\n",
      "    ],\n",
      "    \"M_C_c\": [\n",
      "        \"bact5\",\n",
      "        \"bact6\"\n",
      "    ],\n",
      "    \"M_D_c\": [\n",
      "        \"bact1\"\n",
      "    ],\n",
      "    \"M_E_c\": [\n",
      "        \"bact3\",\n",
      "        \"bact2\"\n",
      "    ],\n",
      "    \"M_F_c\": [\n",
      "        \"bact1\"\n",
      "    ],\n",
      "    \"M_G_c\": [\n",
      "        \"bact3\",\n",
      "        \"bact2\"\n",
      "    ],\n",
      "    \"M_H_c\": [\n",
      "        \"bact3\",\n",
      "        \"bact2\",\n",
      "        \"bact1\"\n",
      "    ],\n",
      "    \"M_K_c\": [\n",
      "        \"bact5\",\n",
      "        \"bact6\"\n",
      "    ],\n",
      "    \"M_N_c\": [\n",
      "        \"bact3\",\n",
      "        \"bact2\"\n",
      "    ],\n",
      "    \"M_R_c\": [\n",
      "        \"bact1\"\n",
      "    ],\n",
      "    \"M_S1_c\": [\n",
      "        \"bact3\"\n",
      "    ],\n",
      "    \"M_S2_c\": [\n",
      "        \"bact2\"\n",
      "    ],\n",
      "    \"M_V_c\": [\n",
      "        \"bact5\"\n",
      "    ],\n",
      "    \"M_X_c\": [\n",
      "        \"bact6\",\n",
      "        \"bact4\"\n",
      "    ]\n",
      "}\n"
     ]
    }
   ],
   "source": [
    "with open('output_folder/community_analysis/rev_cscope.json') as f:\n",
    "    rev_cscope = json.load(f)\n",
    "\n",
    "print(json.dumps(rev_cscope, indent = 4, sort_keys=True))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The matrix of the community scope"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>M_K_c</th>\n",
       "      <th>M_V_c</th>\n",
       "      <th>M_C_c</th>\n",
       "      <th>M_S1_c</th>\n",
       "      <th>M_B_c</th>\n",
       "      <th>M_N_c</th>\n",
       "      <th>M_G_c</th>\n",
       "      <th>M_H_c</th>\n",
       "      <th>M_A_c</th>\n",
       "      <th>M_E_c</th>\n",
       "      <th>M_X_c</th>\n",
       "      <th>M_S2_c</th>\n",
       "      <th>M_R_c</th>\n",
       "      <th>M_D_c</th>\n",
       "      <th>M_F_c</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>bact5</th>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>bact3</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>bact6</th>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>bact2</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>bact4</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>bact1</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       M_K_c  M_V_c  M_C_c  M_S1_c  M_B_c  M_N_c  M_G_c  M_H_c  M_A_c  M_E_c  \\\n",
       "bact5      1      1      1       0      0      0      0      0      0      0   \n",
       "bact3      0      0      0       1      1      1      1      1      1      1   \n",
       "bact6      1      0      1       0      0      0      0      0      0      0   \n",
       "bact2      0      0      0       0      1      1      1      1      1      1   \n",
       "bact4      0      0      0       0      1      0      0      0      0      0   \n",
       "bact1      0      0      0       0      1      0      0      1      1      0   \n",
       "\n",
       "       M_X_c  M_S2_c  M_R_c  M_D_c  M_F_c  \n",
       "bact5      0       0      0      0      0  \n",
       "bact3      0       0      0      0      0  \n",
       "bact6      1       0      0      0      0  \n",
       "bact2      0       1      0      0      0  \n",
       "bact4      1       0      0      0      0  \n",
       "bact1      0       0      1      1      1  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "display(pandas.read_csv('output_folder/community_analysis/rev_cscope.tsv', sep='\\t', index_col='Unnamed: 0'))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Command line from the terminal\n",
    "\n",
    "Most of the time, you will launch m2m from the terminal rather than from the API. \n",
    "The command for community scope computation is the following, it will create all output files:\n",
    "\n",
    "```sh\n",
    "m2m cscope -n ../../test/metabolic_data/tiny_toy/networks/ -s ../../test/metabolic_data/tiny_toy/seeds_community.sbml -o output_folder\n",
    "```\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## `m2m addedvalue`: Cooperation potential\n",
    "\n",
    "This part encompasses the computation of the cooperation potential, i.e. the set of molecules that can be produced in community but not individually given the seeds provided as input.\n",
    "\n",
    "### Cooperation potential formalisation\n",
    "\n",
    "Given individual and community metabolic potentials, the _cooperation potential_ consists in the set of metabolites whose producibility can only occur if several organisms participate in the biosynthesis. `m2m addedvalue` computes the cooperation potential by performing a set difference between the community scope and the union of individual scopes, and produces a SBML file with the resulting metabolites. This list of compounds is inclusive and could comprise false positives not necessitating cooperation for production, but selected due to missing annotations in the initial genomes. One can modify the SBML file accordingly, prior to the following M2M community reduction step.\n",
    "\n",
    "The cooperation potential $\\coopPot(G_1,..,G_n,S)$ of a collection of metabolic networks $\\{G_1..G_n\\}$ is defined by:\n",
    "\n",
    "\n",
    "$\\coopPot(G_1,..,G_n,S) = \\mixedbagScope(G_1,..,G_n,S)  \\setminus  \\bigcup_{i \\in {\\{1..n\\}}} \\Scope(G_i,S).$\n",
    "\n",
    "### An example\n",
    "\n",
    "In the **Figure 4**, the cooperation potential is:\n",
    "\n",
    "$\\coopPot(bact1, bact2, bact3, bact4, bact5, bact6) = \\{A, B, C, D, E, F, G, H, K, N, R, S2, S1, V, X\\} - \\{A, B, D, H, N, R, S2\\} $\n",
    "\n",
    "$\\coopPot(OrgA,OrgB,OrgC,S) = \\{C, E, F, G, H, K, S1, V, X\\} $\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Run the cooperation potential with metage2metabo on the folder of SBML using the seeds files.\n",
    "# This is similar to the command:\n",
    "# m2m addedvalue -n data/community -s data/seeds.sbml -o output_folder\n",
    "from metage2metabo.m2m import community_addedvalue, individual_scope, community_scope\n",
    "\n",
    "networks_path = '../../test/metabolic_data/tiny_toy/networks/'\n",
    "seeds_path = '../../test/metabolic_data/tiny_toy/seeds_community.sbml'\n",
    "output_folder = 'output_folder'\n",
    "\n",
    "iscope_metabolites = individual_scope.iscope(networks_path, seeds_path, output_folder)\n",
    "instance_comscope, cscope_metabolites = community_scope.cscope(networks_path, seeds_path, output_folder)\n",
    "\n",
    "addedvalue = community_addedvalue.addedvalue(iscope_metabolites, cscope_metabolites, output_folder)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The union of individual scopes contains 7 metabolites: M_A_c, M_B_c, M_D_c, M_H_c, M_N_c, M_R_c, M_S2_c.\n",
      "The community scope contains 15 metabolites: M_A_c, M_B_c, M_C_c, M_D_c, M_E_c, M_F_c, M_G_c, M_H_c, M_K_c, M_N_c, M_R_c, M_S1_c, M_S2_c, M_V_c, M_X_c.\n",
      "The cooperation potential contains 8 metabolites: M_C_c, M_E_c, M_F_c, M_G_c, M_K_c, M_S1_c, M_V_c, M_X_c.\n"
     ]
    }
   ],
   "source": [
    "print('The union of individual scopes contains {0} metabolites: {1}.'.format(len(iscope_metabolites), ', '.join(sorted(iscope_metabolites))))\n",
    "print('The community scope contains {0} metabolites: {1}.'.format(len(cscope_metabolites), ', '.join(sorted(cscope_metabolites))))\n",
    "print('The cooperation potential contains {0} metabolites: {1}.'.format(len(addedvalue), ', '.join(sorted(addedvalue))))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Command line from the terminal\n",
    "\n",
    "Most of the time, you will launch m2m from the terminal rather than from the API. \n",
    "The command for added value computation is the following, it will create all output files:\n",
    "\n",
    "```sh\n",
    "m2m addedvalue -n ../../test/metabolic_data/tiny_toy/networks/ -s ../../test/metabolic_data/tiny_toy/seeds_community.sbml -o output_folder\n",
    "```\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## `m2m mincom`: Minimal community selection\n",
    "\n",
    "\n",
    "### Minimal community formalism\n",
    "\n",
    "A minimal community $\\mathcal{C}$ enabling the producibility of a set of targets $T$ from the seeds $S$ is a sub-family of the community $G_1, \\dots, G_n$ which is solution of the following optimisation problem:\n",
    "\n",
    "$SetofMinimalCommunity(G_1 .. G_n, S, T) =  arg \\, min ( size( {\\{G_i..G_j\\}} \\, | \\, \\mathcal{P}_{\\{G_1..G_n\\}} \\subset  \\Sigma_{\\{G_i..G_j\\}}(S) )$\n",
    "\n",
    "$$\\begin{array}{lll}\n",
    " \\qquad \\qquad  & \n",
    " \\displaystyle{\n",
    " \\minimize_{ \\{G_{i_1}.. G_{i_L}\\}  \\  \\subset \\{G_{1}.. G_{N}\\} }\n",
    " }\n",
    "  & \\mbox{size} (\\{G_{i_1}.. G_{i_L}\\})   \\\\ \n",
    "& \\mbox{subject to }  &   T \\subset  \\mixedbagScope(G_{i_1} .. G_{i_L}),S). \n",
    "\\end{array}       \n",
    "$$\n",
    "\n",
    "### Example\n",
    "\n",
    "In this example, we have 8 species. In order to produce the targets from the seeds, there are 4 minimal communities of size of 3 organisms.\n",
    "\n",
    "The yellow organism occurs in all minimal communities, it is an essential symbiont.\n",
    "Green, orange, cyan and purple organisms occur in some minimal communities but not in all, they are alternatives symbionts.\n",
    "\n",
    "<p>\n",
    "    <img src='images/key_species.png' width='40%'>\n",
    "    <center><em>Figure 6: Key species</em></center>\n",
    "</p>\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Key species\n",
    "\n",
    "Many minimal communities are expected to be equivalent for a given metabolic objective but their enumeration can be computationally costly.\n",
    "\n",
    "We define *key species* which are organisms occurring in at least one community among all the optimal ones. Key species can be further distinguished into *essential symbionts* and *alternative symbionts*. The former occur in every minimal community whereas the latter occur only in some minimal communities. More precisely, the key species $\\ks(G_1 .. G_n, S,T)$, the essential symbionts $\\es(G_1 .. G_n, S,T)$, and the alternative symbionts $\\as(G_1 .. G_n, S,T)$ associated to a set of metabolic networks, seeds $S$ and a set of target metabolites $T$ are defined by:\n",
    "\n",
    "\n",
    "\\begin{align*}\n",
    "    \\ks(G_1 .. G_n, S,T) &= \\{ G \\mid \\exists {\\mathcal C} \\in \\mincom(G_1 .. G_n, S, T), \\, G \\in {\\mathcal C}\\}. \\\\\n",
    "    \\es(G_1 .. G_n, S,T) &= \\{ G \\mid \\forall {\\mathcal C} \\in \\mincom(G_1 .. G_n, S, T), \\, G \\in {\\mathcal C}\\}. \\\\\n",
    "    \\as(G_1 .. G_n, S,T) &= \\ks(G_1 .. G_n, S,T) \\setminus \\es(G_1 .. G_n, S,T).\n",
    "\\end{align*}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/home/abelcour/Documents/Softwares/git/metage2metabo/tutorials/method_tutorial/output_folder/community_analysis/miscoto_6dr7c6nv.lp\n",
      "/home/abelcour/.local/lib/python3.10/site-packages/miscoto/encodings/community_soup.lp\n"
     ]
    }
   ],
   "source": [
    "# Run the minimal community with metage2metabo on the folder of SBML using the seeds files.\n",
    "# This is similar to the command:\n",
    "#m2m metacom -n data/community -s data/seeds.sbml -o output_folder\n",
    "from metage2metabo.m2m import community_addedvalue, individual_scope, community_scope, minimal_community\n",
    "from metage2metabo import sbml_management\n",
    "\n",
    "networks_path = '../../test/metabolic_data/tiny_toy/networks/'\n",
    "seeds_path = '../../test/metabolic_data/tiny_toy/seeds_community.sbml'\n",
    "output_folder = 'output_folder'\n",
    "target_path = '../../test/metabolic_data/tiny_toy/targets_community.sbml'\n",
    "\n",
    "iscope_metabolites = individual_scope.iscope(networks_path, seeds_path, output_folder)\n",
    "instance_comscope, cscope_metabolites = community_scope.cscope(networks_path, seeds_path, output_folder)\n",
    "addedvalue = community_addedvalue.addedvalue(iscope_metabolites, cscope_metabolites, output_folder)\n",
    "\n",
    "instance_mincom = community_scope.instance_community(networks_path, seeds_path, output_folder, target_path)\n",
    "\n",
    "# The path contains the instance for Answer Set Programming computation of minimal communities.\n",
    "print(instance_mincom)\n",
    "\n",
    "# let's run mincom\n",
    "\n",
    "minimal_community.mincom(instance_mincom, seeds_path, target_path, output_folder)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The mincom command creates the output `community_analysis/mincom.json`. \n",
    "\n",
    "When called with the command line, logs provide an explanation of the results:\n",
    "\n",
    "```\n",
    "m2m mincom -n ../../test/metabolic_data/tiny_toy/networks -s ../../test/metabolic_data/tiny_toy/seeds_community.sbml -t ../../test/metabolic_data/tiny_toy/targets_community.sbml -o output_folder\n",
    "######### Creating metabolic instance for the whole community #########\n",
    "Created temporary instance file in output_folder/community_analysis/miscoto__3lz5o7z.lp\n",
    "\n",
    "###############################################\n",
    "#                                             #\n",
    "#         Minimal community selection         #\n",
    "#                                             #\n",
    "###############################################\n",
    "\n",
    "Running minimal community selection\n",
    "/Users/cfrioux/.pyenv/versions/metage2metabo/lib/python3.10/site-packages/miscoto/encodings/community_soup.lp\n",
    "\n",
    "In the initial and minimal communities 3 targets are producible and 1 remain unproducible.\n",
    "\n",
    "3 producible targets:\n",
    "M_F_c\n",
    "M_C_c\n",
    "M_H_c\n",
    "\n",
    "1 still unproducible targets:\n",
    "M_foo_c\n",
    "\n",
    "Minimal communities are available in output_folder/community_analysis/mincom.json\n",
    "\n",
    "######### One minimal community #########\n",
    "# One minimal community enabling the producibility of the target metabolites given as inputs\n",
    "Minimal number of bacteria in communities => 3\n",
    "\n",
    "bact5\n",
    "bact1\n",
    "bact3\n",
    "######### Key species: Union of minimal communities #########\n",
    "# Bacteria occurring in at least one minimal community enabling the producibility of the target metabolites given as inputs\n",
    "Number of key species => 5\n",
    "\n",
    "bact1\n",
    "bact5\n",
    "bact2\n",
    "bact6\n",
    "bact3\n",
    "######### Essential symbionts: Intersection of minimal communities #########\n",
    "# Bacteria occurring in ALL minimal communities enabling the producibility of the target metabolites given as inputs\n",
    "Number of essential symbionts => 1\n",
    "\n",
    "bact1\n",
    "######### Alternative symbionts: Difference between Union and Intersection #########\n",
    "# Bacteria occurring in at least one minimal community but not all minimal communities enabling the producibility of the target metabolites given as inputs\n",
    "Number of alternative symbionts => 4\n",
    "\n",
    "bact2\n",
    "bact3\n",
    "bact6\n",
    "bact5\n",
    "\n",
    "--- Mincom runtime 4.99 seconds ---\n",
    "\n",
    "--- Total runtime 5.06 seconds ---\n",
    "\n",
    "```\n",
    "\n",
    "The figure below illustrates the one solution that mincom provides, with the underlying minimal metabolic exchanges that would be necessary (can be computed with the `miscoto` package, a dependency of m2m).\n",
    "\n",
    "<p>\n",
    "    <img src='images/tiny_toy_mincom_onesol.png' width='60%'>\n",
    "    <center><em>Figure 6: One minimal solution enabling the production of the 3 producible targets</em></center>\n",
    "</p>\n",
    "\n",
    "One target cannot be produced. It is the $foo$ metabolite that actually occurs in none of the metabolic networks. \n",
    "\n",
    "More information can be computed regarding target producibility.\n",
    "It is done automatically when calling `m2m metacom`, e.g. `m2m metacom -n ../../test/metabolic_data/tiny_toy/networks -s ../../test/metabolic_data/tiny_toy/seeds_community.sbml -t ../../test/metabolic_data/tiny_toy/targets_community.sbml -o output_folder` and is stored in `output_folder/producibility_targets.json`.\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The file `producibility_targets.json` provides such information, stored as a dictionnary:\n",
    "\n",
    "- Producible targets\n",
    "  ```\n",
    "    \"producible\": [\n",
    "        \"M_H_c\",\n",
    "        \"M_C_c\",\n",
    "        \"M_F_c\"\n",
    "  ```\n",
    "- Unproducible targets\n",
    "  ```\n",
    "      \"unproducible\": [\n",
    "        \"M_foo_c\"\n",
    "    ],\n",
    "  ```\n",
    "- Targets that can be produced by individual species, i.e. do not need cooperation events, together with their producers:\n",
    "  ```\n",
    "    \"indiv_producible\": [\n",
    "        \"M_H_c\"\n",
    "    ],\n",
    "    \"individual_producers\": {\n",
    "        \"M_H_c\": [\n",
    "            \"bact1\"\n",
    "        ]\n",
    "    },\n",
    "  ```\n",
    "- Targets that requires cooperation or cross-feeding for their production, together with their final producers\n",
    "   ```\n",
    "    \"com_only_producers\": {\n",
    "        \"M_H_c\": [\n",
    "            \"bact3\",\n",
    "            \"bact2\"\n",
    "        ],\n",
    "        \"M_F_c\": [\n",
    "            \"bact1\"\n",
    "        ],\n",
    "        \"M_C_c\": [\n",
    "            \"bact6\",\n",
    "            \"bact5\"\n",
    "        ]\n",
    "    },\n",
    "   ```\n",
    "- Key species\n",
    "  ```\n",
    "    \"key_species\": [\n",
    "        \"bact3\",\n",
    "        \"bact2\",\n",
    "        \"bact1\",\n",
    "        \"bact6\",\n",
    "        \"bact5\"\n",
    "    ],\n",
    "  ```\n",
    "- Targets producible in the minimal community\n",
    "  ```\n",
    "     \"mincom_producible\": [\n",
    "        \"M_H_c\",\n",
    "        \"M_C_c\",\n",
    "        \"M_F_c\"\n",
    "    ],\n",
    "  ```\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Please note that for computational efficiency, only one solution, the key species and essential/alternative symbionts are computed. Enumerating all equivalent minimal communities takes more time but can be computed. That is the purpose of `m2m analysis`."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Command line from the terminal\n",
    "\n",
    "Most of the time, you will launch m2m from the terminal rather than from the API. \n",
    "The command for minimal community computation is the following, it will create all output files:\n",
    "\n",
    "```sh\n",
    "m2m mincom -n ../../test/metabolic_data/tiny_toy/networks/ -s ../../test/metabolic_data/tiny_toy/seeds_community.sbml -t ../../test/metabolic_data/tiny_toy/targets_community.sbml -o output_folder\n",
    "```\n",
    "\n",
    "`m2m metacom` runs the complete analysis from the individual scope to the computation of target producibility after minimal community selection. \n",
    "We advise to use directly this command to explore the metabolic potential of communities. \n",
    "\n",
    "```sh\n",
    "m2m metacom -n ../../test/metabolic_data/tiny_toy/networks/ -s ../../test/metabolic_data/tiny_toy/seeds_community.sbml -t ../../test/metabolic_data/tiny_toy/targets_community.sbml -o output_folder\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## `m2m_analysis`: Visualisation of minimal communities\n",
    "\n",
    "To complete the analysis of minimal communities and visualise the results, we will use the m2m_analysis part of m2m. This part is more computational heavy due to the fact that it will enumerate all minimal communities (especially it can need more RAM).\n",
    "\n",
    "Then using the results of enumeration, it will try to create visualisation of the minimal communities."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "from metage2metabo.m2m_analysis import m2m_analysis_workflow\n",
    "\n",
    "# we need a target file in which all molecules can be producible. \n",
    "# foo is not producible, so we'll use another target file that does not contain it\n",
    "\n",
    "networks_path = '../../test/metabolic_data/tiny_toy/networks/'\n",
    "seeds_path = '../../test/metabolic_data/tiny_toy/seeds_community.sbml'\n",
    "output_folder = 'output_folder'\n",
    "target_path = '../../test/metabolic_data/tiny_toy/targets_community_allprod.sbml'\n",
    "\n",
    "m2m_analysis_workflow.run_analysis_workflow(networks_path, target_path, seeds_path, output_folder, None, None)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "As seen in part `m2m mincom`, we have mminimal communities composed of 3 organisms. Now using `m2m_analysis`, we will enumerate these different minimal communties.\n",
    "\n",
    "Minimal communities are described in `output_folder/json/targets_community_allprod.json`.\n",
    "There are 4 minimal communities:\n",
    "\n",
    "- **`bact1`** X `bact3` x `bact6`\n",
    "- **`bact1`** X `bact3` x `bact5`\n",
    "- **`bact1`** X `bact2` x `bact6`\n",
    "- **`bact1`** X `bact2` x `bact5`\n",
    "\n",
    "The 4 communities are illustrated below:\n",
    "\n",
    "<p>\n",
    "    <img src='images/tiny_toy_enum_sol.png' width='60%'>\n",
    "    <center><em>Figure 7: All minimal community solutions</em></center>\n",
    "</p>\n",
    "\n",
    "Then, `m2m_analysis` will compute the solution graph. This graph's nodes correspond to organism from the minimal community. An edge is drawn between two nodes if the two organisms occur in the same minimal community.\n",
    "\n",
    "This graph can be very dense and hard to visualise if a lot of minimal solutions exist. \n",
    "It is stored in `output_folder/gml/targets_community_allprod.gml\n",
    "\n",
    "<p>\n",
    "    <img src='images/tiny_toy_association_graph.png' width='40%'>\n",
    "    <center><em>Figure 7: All minimal community solutions</em></center>\n",
    "</p>\n",
    "\n",
    "Then m2m_analysis will compress the solution graph into a [power graph](https://en.wikipedia.org/wiki/Power_graph_analysis). To do this, m2m_analysis uses the `powergrasp` package, which will search for specific pattern (like hubs or cliques) and compress them into power nodes and power edges. Here, as node `bact1` is a hub in the graph (as connected to the other nodes of the graph). We notice that `bact2` and `bact3` have the same roles in minimal communities, likewise for `bact5` and `bact6`. Each pair will belong to a power node, indicating they can replace each other, i.e. a minimal community holds one of the two pair elements. \n",
    "\n",
    "`bact1` occurs in all the minimal communities, it is an essential symbiont (in dark pink).\n",
    "\n",
    "`bact2`, `bact3`, `bact5` and `bact6` occur in at least one minimal community but not all, they are alternative symbionts (in blue).\n",
    "\n",
    "<p>\n",
    "    <img src='images/tiny_toy_powergraph.png' width='40%'>\n",
    "    <center><em>Figue 9: Powergraph</em></center>\n",
    "</p>\n",
    "\n",
    "The power graph information is stored in `output_folder/bbl`. It can be visualised as an html page `html/targets_community_allprod_powergraph.html`."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Command line from the terminal\n",
    "\n",
    "Most of the time, you will launch m2m_analysis from the terminal rather than from the API. \n",
    "The command is the following, it will create all output files:\n",
    "\n",
    "```sh\n",
    "m2m_analysis workflow -n ../../test/metabolic_data/tiny_toy/networks/ -s ../../test/metabolic_data/tiny_toy/seeds_community.sbml -t ../../test/metabolic_data/tiny_toy/targets_community_allprod.sbml -o output_folder\n",
    "```\n",
    "\n",
    "Other options are available, for instance for creating the svg image of the power graph, or taking into account the taxonomy of the community members. Take a look at the command line help `m2m_analysis workflow --help` and the documentation.\n",
    "\n",
    "Here the whole workflow is run. You can launch parts of the workflow, take a look at the documentation and the help of the tool using `m2m_analysis -h`."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The command produces the following logs:\n",
    "\n",
    "```\n",
    "###############################################\n",
    "#                                             #\n",
    "#      Enumeration of minimal communities     #\n",
    "#                                             #\n",
    "###############################################\n",
    "\n",
    "######### Enumeration of solution for: targets_community_allprod #########\n",
    "/shared/Softwares/git/miscoto/miscoto/encodings/community_soup.lp\n",
    "######### Enumeration of minimal communities #########\n",
    "4 minimal communities (each containing 3 species) producing the target metabolites\n",
    "######### Key species: Union of minimal communities #########\n",
    "# Bacteria occurring in at least one minimal community enabling the producibility of the target metabolites given as inputs\n",
    "Key species = 5\n",
    "bact1\n",
    "bact5\n",
    "bact2\n",
    "bact3\n",
    "bact6\n",
    "######### Essential symbionts: Intersection of minimal communities #########\n",
    "# Bacteria occurring in ALL minimal community enabling the producibility of the target metabolites given as inputs\n",
    "Essential symbionts = 1\n",
    "bact1\n",
    "######### Alternative symbionts: Difference between Union and Intersection #########\n",
    "# Bacteria occurring in at least one minimal community but not all minimal community enabling the producibility of the target metabolites given as inputs\n",
    "Alternative symbionts = 4\n",
    "bact3\n",
    "bact6\n",
    "bact2\n",
    "bact5\n",
    "--- Enumeration runtime 0.79 seconds ---\n",
    "\n",
    "\n",
    "###############################################\n",
    "#                                             #\n",
    "#         Solution graph creation             #\n",
    "#                                             #\n",
    "###############################################\n",
    "\n",
    "######### Graph of targets_community_allprod #########\n",
    "Number of nodes: 5\n",
    "Number of edges: 8\n",
    "--- Graph runtime 0.01 seconds ---\n",
    "\n",
    "\n",
    "###############################################\n",
    "#                                             #\n",
    "#  Compression and visualisation of graph     #\n",
    "#                                             #\n",
    "###############################################\n",
    "\n",
    "######### Graph compression: targets_community_allprod #########\n",
    "Number of powernodes: 3\n",
    "Number of poweredges: 2\n",
    "Compression runtime 1.43 seconds ---\n",
    "\n",
    "######### Test powergraph heuristics: targets_community_allprod #########\n",
    "Same combinations between theorical (4) and solution (4)\n",
    "The powergraph seems to be an optimal representation of the solutions.\n",
    "\n",
    "Same combinations between theorical (4) and the enumeration of solutions by m2m_analysis  (4)\n",
    "The powergraph seems to be an optimal representation of the solutions.\n",
    "\n",
    "Same number of solution between the computed combinations from powernodes (4) and the enumeration of solutions by m2m_analysis (4)\n",
    "But this does not indicate that the powernodes are an optimal representation but that they contain the solution.\n",
    "\n",
    "It seems that there are no heuristics in powergraph so it could be possible to create a boolean equation.\n",
    "######### Boolean equation of minimal communities #########\n",
    "The boolean equation can only be created for simple case (without too many combinations).\n",
    "Boolean equation seems good, as it has the same combinations (4) than the one from enumeration (4).\n",
    "Boolean equation: \n",
    "(( bact1 ) & \n",
    "( bact6 | bact5 ) & \n",
    "( bact3 | bact2 ) )\n",
    "######### PowerGraph visualization: targets_community_allprod #########\n",
    "Creation of the powergraph website accessible at output_folder/html/targets_community_allprod\n",
    "--- Powergraph runtime 1.44 seconds ---\n",
    "\n",
    "--- m2m_analysis runtime 2.24 seconds ---\n",
    "\n",
    "--- Total runtime 2.24 seconds ---\n",
    "\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Graph compression can used an heuristics if the combination are too complex. If this is the cas, `m2m_analysis` will write warning in the log during the `Test powergraph heuristics:` step.\n",
    "\n",
    "Furthermore, this will prevent the creation of a boolean equation summarizing the minimal communities.\n",
    "\n",
    "But if the combination are simple a boolean equation can be created. The powergraph is then sumarized into a boolean equation:\n",
    "\n",
    "`(( bact1 ) & ( bact3 | bact2 ) & ( bact6 | bact5 ) )`\n",
    "\n",
    "`&` corresponds to `AND` logical operator. `|` corresponds to `OR` logical operator. If an organism is alone between parenthesis, it is an essential symbiont. If it is within parenthesis with other organisms and separated by `|`, it is an alternative symbiont.\n",
    "\n",
    "It reads as follow: to produce the target metabolites we need `bact1` and either `bact3 or bact2` and either `bact6 or bact5`."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Tutorial on real data\n",
    "\n",
    "For this tutorial, we used a subset of the data used in the article of M2M (avaible in the [Github repository](https://github.com/AuReMe/metage2metabo/tree/master/test/metabolic_data)).\n",
    "\n",
    "- `toy_bact`: 17 metabolic networks\n",
    "- `seeds_toy.sbml`: 93 metabolites (classical diet for the gut microbiota, EU average from the VMH resource and a small number of currency metabolites).\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Average individual scope: 239.05882352941177\n",
      "The community scope contains 698 metabolites.\n",
      "Cooperation potential: 122 metabolites.\n",
      "\n",
      "\n",
      "12 Essential symbionts: GCA_003437715,GCA_003437195,GCA_003438055,GCA_003437815,GCA_003437885,GCA_003437295,GCA_003437055,GCA_003437255,GCA_003437375,GCA_003437665,GCA_003437905,GCA_003437595\n",
      "5 Alternative symbionts: GCA_003437345,GCA_003437325,GCA_003437175,GCA_003437785,GCA_003437945\n",
      "\n",
      "\n",
      "Boolean equation: (( GCA_003437715 ) & ( GCA_003437945 | GCA_003437325 | GCA_003437175 | GCA_003437345 | GCA_003437785 ) & ( GCA_003437195 ) & ( GCA_003438055 ) & ( GCA_003437815 ) & ( GCA_003437885 ) & ( GCA_003437295 ) & ( GCA_003437055 ) & ( GCA_003437255 ) & ( GCA_003437375 ) & ( GCA_003437665 ) & ( GCA_003437905 ) & ( GCA_003437595 ) )\n"
     ]
    }
   ],
   "source": [
    "# Instanciate the logger\n",
    "import logging\n",
    "import statistics\n",
    "import json\n",
    "import os\n",
    "\n",
    "logging.basicConfig()\n",
    "logging.setLevel = logging.CRITICAL\n",
    "# Run the individual scope with metage2metabo on the folder of SBML using the seeds files.\n",
    "# This is similar to the command:\n",
    "# m2m iscope -n data/community -s data/seeds.sbml -o output_folder\n",
    "from metage2metabo.m2m import individual_scope\n",
    "\n",
    "networks_path = '../../test/metabolic_data/toy_bact'\n",
    "seeds_path = '../../test/metabolic_data/seeds_toy.sbml'\n",
    "output_folder = 'tutorial_output_folder'\n",
    "indiv_output = os.path.join(output_folder, 'indiv_scopes')\n",
    "indiv_scope_json = os.path.join(indiv_output, 'indiv_scopes.json')\n",
    "key_species_file = os.path.join(output_folder, 'key_species.json')\n",
    "target_path = '../../test/metabolic_data/targets_toy.sbml'\n",
    "\n",
    "individual_scope.iscope(networks_path, seeds_path, output_folder)\n",
    "\n",
    "# Read the result\n",
    "with open(indiv_scope_json) as iscope_output:\n",
    "    iscope_json_data = json.load(iscope_output)\n",
    "\n",
    "metabolites = []\n",
    "union_metabolites = set()\n",
    "for species in iscope_json_data:\n",
    "    metabolites.append(len(iscope_json_data[species]))\n",
    "    union_metabolites.update(iscope_json_data[species])\n",
    "\n",
    "print('Average individual scope: {0}'.format(statistics.mean(metabolites)))\n",
    "\n",
    "# Run the community scope with metage2metabo on the folder of SBML using the seeds files.\n",
    "# This is similar to the command: m2m cscope -n toy_bact -s seeds_toy.sbml -o output_folder\n",
    "from metage2metabo.m2m import community_scope\n",
    "\n",
    "instance_path, network_scopes = community_scope.cscope(networks_path, seeds_path, output_folder)\n",
    "\n",
    "print('The community scope contains {0} metabolites.'.format(len(network_scopes)))\n",
    "\n",
    "print('Cooperation potential: {0} metabolites.'.format(len(set(network_scopes)-union_metabolites)))\n",
    "# Run the minimal community with metage2metabo on the folder of SBML using the seeds files.\n",
    "# This is similar to the command: m2m metacom -n toy_bact -s seeds_toy.sbml -o output_folder\n",
    "from metage2metabo.m2m import community_addedvalue, individual_scope, community_scope\n",
    "from metage2metabo import sbml_management\n",
    "\n",
    "iscope_metabolites = individual_scope.iscope(networks_path, seeds_path, output_folder)\n",
    "instance_comscope, cscope_metabolites = community_scope.cscope(networks_path, seeds_path, output_folder)\n",
    "addedvalue = community_addedvalue.addedvalue(iscope_metabolites, cscope_metabolites, output_folder)\n",
    "\n",
    "if len(addedvalue):\n",
    "    sbml_management.create_species_sbml(addedvalue, target_path)\n",
    "instance_mincom = community_scope.instance_community(networks_path, seeds_path, output_folder, target_path)\n",
    "\n",
    "from metage2metabo.m2m_analysis import m2m_analysis_workflow\n",
    "\n",
    "m2m_analysis_workflow.run_analysis_workflow(networks_path, target_path, seeds_path, output_folder, None, None)\n",
    "\n",
    "with open(key_species_file) as json_key_species:\n",
    "    key_species_json_data = json.load(json_key_species)\n",
    "\n",
    "nb_essential_symbionts = len(key_species_json_data['targets_toy']['essential_symbionts']['data'])\n",
    "essential_symbionts = ','.join(key_species_json_data['targets_toy']['essential_symbionts']['data'])\n",
    "nb_alternative_symbionts = len(key_species_json_data['targets_toy']['alternative_symbionts']['data'])\n",
    "alternative_symbiotns = ','.join(key_species_json_data['targets_toy']['alternative_symbionts']['data'])\n",
    "print('\\n')\n",
    "print('{0} Essential symbionts: {1}'.format(nb_essential_symbionts, essential_symbionts))\n",
    "print('{0} Alternative symbionts: {1}'.format(nb_alternative_symbionts, alternative_symbiotns))\n",
    "\n",
    "boolean_equation_json_file = os.path.join(output_folder, 'minimal_equation', 'targets_toy', 'boolean_equation.json')\n",
    "with open(boolean_equation_json_file) as open_boolean_equation_json:\n",
    "    boolean_equation_data = json.load(open_boolean_equation_json)\n",
    "\n",
    "print('\\n')\n",
    "print('Boolean equation: {0}'.format(boolean_equation_data['boolean_equation']))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "If you want more information on the output files:\n",
    "\n",
    "- result files of m2m are described [here](https://metage2metabo.readthedocs.io/en/latest/output.html).\n",
    "- result files of m2m_analysis are explained [here](https://metage2metabo.readthedocs.io/en/latest/m2m_analysis.html#m2m-analysis-output-files)."
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
